javaslang/src/main/java/javaslang/λ.java

```diff
@@ -74,5 +75,17 @@ default boolean isMemoized() {
      * Zero Abstract Method (ZAM) interface for marking functions as memoized using intersection types.
      */
     interface Memoized {
+        static <T extends Tuple, R> R of(Map<T, R> cache, T key, Function1<T, R> tupled) {
```

Comment:
I think this is a good solution, we can do that. I currently have not a better solution for the synching problem. It is hard to test.

I'm not that familiar with proper cache invalidation (Weak/Soft-Refs) but we need to keep the tuples in the cache as long as the memoized function is referenced.